### PR TITLE
Spark: Fix UnresolvedException for some filters in rewrite_data_files procedure

### LINKED ISSUE #3756: In-filter throws exception for rewrite_data_files procedure.
call stack:
```
Invalid call to dataType on unresolved object
org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to dataType on unresolved object
  at app//org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute.dataType(unresolved.scala:160)
  at app//org.apache.spark.sql.execution.datasources.DataSourceStrategy$.translateLeafNodeFilter(DataSourceStrategy.scala:533)
  at app//org.apache.spark.sql.execution.datasources.DataSourceStrategy$.translateFilterWithMapping(DataSourceStrategy.scala:614)
  at app//org.apache.spark.sql.execution.datasources.DataSourceStrategy$.translateFilter(DataSourceStrategy.scala:565)
  at app//org.apache.spark.sql.execution.datasources.SparkExpressionConverter$.convertToIcebergExpression(SparkExpressionConverter.scala:31)
  at app//org.apache.spark.sql.execution.datasources.SparkExpressionConverter.convertToIcebergExpression(SparkExpressionConverter.scala)
  at app//org.apache.iceberg.spark.procedures.RewriteDataFilesProcedure.checkAndApplyFilter(RewriteDataFilesProcedure.java:125)
  at app//org.apache.iceberg.spark.procedures.RewriteDataFilesProcedure.lambda$call$0(RewriteDataFilesProcedure.java:112)
```

Cause: 
sqlParser.parseExpression(where) gives unresolved expression. Most of the filters doesn't need it to be resolved. Some special filter like In-filter needs resolved type because based on datatype it creates HashSet.

Solution:
Use a resolved expression.
I found a hacky sparkish way to achieve it. Will raise a PR for the same.

cc: @RussellSpitzer , @jackye1995

### PR DESCRIPTION
**Pull request by ajantha-bhat**
Some filters fails with exception mentioned in #3756.
Hence use the spark resolved expressions in rewrite_data_files procedure

Fixes #3756

### CODE REVIEW

**Comment by ajantha-bhat** [comment]
cc: @RussellSpitzer , @jackye1995

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java [hunk 1]
```diff
@@ -109,19 +110,23 @@ public StructType outputType() {
...
 
       return toOutputRows(result);
     });
   }
 
-  private RewriteDataFiles checkAndApplyFilter(RewriteDataFiles action, String where) {
+  private RewriteDataFiles checkAndApplyFilter(RewriteDataFiles action, String where, String tableName) {
     if (where != null) {
+      // Add a dummy prefix linking to the table to collect the resolved spark expression from analyzed plan.
+      String prefix = String.format("SELECT count(*) from %s where ", tableName);
```
Comment (huaxingao) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java:124]: I think there is no need to have an aggregate function here.  A simple query `SELECT 1 from %s where` is sufficient.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java [hunk 2]
```diff
@@ -109,19 +110,23 @@ public StructType outputType() {
...
       ParserInterface sqlParser = spark().sessionState().sqlParser();
       try {
-        Expression expression = sqlParser.parseExpression(where);
+        LogicalPlan logicalPlan = sqlParser.parsePlan(prefix + where);
+        Expression expression = SparkExpressionConverter.collectSparkExpression(
+            Dataset.ofRows(spark(), logicalPlan).queryExecution().analyzed());
```
Comment (huaxingao) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java:129]: Instead of calling
 ```
Dataset.ofRows(spark(), logicalPlan).queryExecution().analyzed()
```
we might want to directly call
 ```
spark().sessionState().executePlan(logicalPlan, CommandExecutionMode.ALL()).analyzed()
```

**Comment by jackye1995** [comment]
Thanks. I think this is short enough to fix, could you directly do it across all versions affected?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java (hunk 2 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java]: ok. 
@huaxingao : I have moved this logic into SparkExpressionConverter#collectResolvedSparkExpression.
So, it can be reused in your #3745 PR.

**Comment by ajantha-bhat** [comment]
> Thanks. I think this is short enough to fix, could you directly do it across all versions affected?

@jackye1995 : I will do that once PR is approved. Else I might have to rework on multiple files if I get comments.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java [hunk 3]
```diff
@@ -109,21 +106,20 @@ public StructType outputType() {
...
-  private RewriteDataFiles checkAndApplyFilter(RewriteDataFiles action, String where) {
+  private RewriteDataFiles checkAndApplyFilter(RewriteDataFiles action, String where, String tableName) {
     if (where != null) {
-      ParserInterface sqlParser = spark().sessionState().sqlParser();
       try {
-        Expression expression = sqlParser.parseExpression(where);
-        return action.filter(SparkExpressionConverter.convertToIcebergExpression(expression));
-      } catch (ParseException e) {
+        return action.filter(SparkExpressionConverter.convertToIcebergExpression(
+            SparkExpressionConverter.collectResolvedSparkExpression(spark(), tableName, where)));
```
Comment (jackye1995) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java:121]:
```suggestion
        Expression expression = SparkExpressionConverter.collectResolvedSparkExpression(spark(), tableName, where);
        return action.filter(SparkExpressionConverter.convertToIcebergExpression(expression));
```

File: spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala [hunk 4]
```diff
@@ -30,4 +34,18 @@ object SparkExpressionConverter {
     // But these two conversions already exist and well tested. So, we are going with this approach.
     SparkFilters.convert(DataSourceStrategy.translateFilter(sparkExpression, supportNestedPredicatePushdown = true).get)
   }
+
+  @throws[AnalysisException]
+  def collectResolvedSparkExpression(session: SparkSession, tableName: String, where: String): Expression = {
+    var expression:Expression = null
+    // Add a dummy prefix linking to the table to collect the resolved spark expression from analyzed plan.
+    val prefix = String.format("SELECT 42 from %s where ", tableName)
```
Comment (jackye1995) [inline spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:42]: nit: L42-43 can be combined, `prefix` is never used after L43.

File: spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala [hunk 5]
```diff
@@ -30,4 +34,18 @@ object SparkExpressionConverter {
     // But these two conversions already exist and well tested. So, we are going with this approach.
     SparkFilters.convert(DataSourceStrategy.translateFilter(sparkExpression, supportNestedPredicatePushdown = true).get)
   }
+
+  @throws[AnalysisException]
+  def collectResolvedSparkExpression(session: SparkSession, tableName: String, where: String): Expression = {
```
Comment (jackye1995) [inline spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:39]: yeah it's hacky but I cannot think of any other better way, @RussellSpitzer any thoughts?

Comment (RussellSpitzer) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:39]: I mean Ideally we could get all this into Spark so Spark could do these conversion for us and just give us filters, but I don't think there are any ways around this at the moment.

Comment (RussellSpitzer) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:39]: Well one way I did do this previously, is I just made the full DatasourceRelationNode and just pulled the filters out of that. I'm not sure that's any cleaner than this

Comment (jackye1995) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:39]: yeah I think this would actually be cleaner than using DatasourceRelationNode, I am good with this impl, please let us know if you have any additional concern.

Comment (RussellSpitzer) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:39]: Yep, lgtm. I think we may want to use similar logic in our expire snapshots procedure as well to allow functions and other Spark things

File: spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala [hunk 6]
```diff
@@ -30,4 +34,18 @@ object SparkExpressionConverter {
...
+    var expression:Expression = null
+    // Add a dummy prefix linking to the table to collect the resolved spark expression from analyzed plan.
+    val prefix = String.format("SELECT 42 from %s where ", tableName)
+    val logicalPlan = session.sessionState.sqlParser.parsePlan(prefix + where)
+    val analyzedLogicalPlan = session.sessionState.executePlan(logicalPlan, CommandExecutionMode.ALL).analyzed
+    analyzedLogicalPlan.collectFirst {
+      case filter: Filter =>
+        expression = filter.expressions.head
+    }
+    expression
```
Comment (RussellSpitzer) [inline spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:49]: One question here, since we are pulling out the relation node here to get these filters .. do we still need
```convertToIcebergExpression```?

Since we basically can just call
```
SparkFilters.convert(analyzedLogicalPlan.collectFirst { case filter: Filter =>  expression = filter.expressions.head }) 
```
?

Also is ok to just use collectFirst? Don't we need all possible Filter expressions? I may have forgotten how this works.

File: spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala (hunk 5 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:39]: ok, I will check expire snapshots procedure in separate PR. It is currently not the scope of this PR.

File: spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala (hunk 4 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:42]: I kept it intentionally. so, I can add a comment about prefix and easy to read

File: spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala (hunk 6 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:49]: a. we can't use SparkFilters.convert, as Filter in plan is `org.apache.spark.sql.catalyst.plans.logical.Filter` and Filter in SparkFilters.convert is `org.apache.spark.sql.sources.Filter`

b. I think we can use collectFirst as the plan is fixed always as query is fixed. It is simple Project + Filter + scan operators in the pan.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java (hunk 3 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java:121]: Here I thought expression is used only to pass below. So, why to has a variable for that. But  I will separate it out. It is more readable.

Comment (ajantha-bhat) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java]: done

**Comment by ajantha-bhat** [comment]
@jackye1995 , @RussellSpitzer @huaxingao : I have addressed the comments. Please have a look at it again

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java [hunk 7]
```diff
@@ -174,6 +174,27 @@ public void testRewriteDataFilesWithFilterOnPartitionTable() {
     assertEquals("Data after compaction should not change", expectedRecords, actualRecords);
   }
 
+  @Test
+  public void testRewriteDataFilesWithInFilterOnPartitionTable() {
```
Comment (RussellSpitzer) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:178]: Lets add in some tests for some of the other possible Filters, I think it would even be ok to do all of them in one test but it may make sense to separate them out for readability. 

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala

Comment (ajantha-bhat) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:178]: Do we need check for all data type aswell ? Then it may become 15 (type of filters) * 10 (type of data type) scenarios

Comment (RussellSpitzer) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:178]: I would just put all filters in the same query, or maybe group into 3 queries, I don't think data types is important here. At least from my perspective the difficult here is in resolving the operator portion, not the literal

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java [hunk 8]
```diff
@@ -174,6 +174,80 @@ public void testRewriteDataFilesWithFilterOnPartitionTable() {
...
+    sql("CALL %s.system.rewrite_data_files(table => '%s'," +
+            " where => 'c1 = 3 and c2 = \"bar\"')", catalogName, tableIdent);
+    // Or
+    sql("CALL %s.system.rewrite_data_files(table => '%s'," +
+            " where => 'c1 = 3 or c1 = 5')", catalogName, tableIdent);
+    // Not
+    sql("CALL %s.system.rewrite_data_files(table => '%s'," +
+            " where => 'c1 not in (1,2)')", catalogName, tableIdent);
+    // StringStartsWith
+    sql("CALL %s.system.rewrite_data_files(table => '%s'," +
```
Comment (ajantha-bhat) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:241]: @RussellSpitzer : I added more testcase.
For, StringStartsWith I don't know what is the other syntax. I tried it as "like %car". 
So, it is coming as LIKE filter instead of  StringStartsWith.

a) What is the SQL syntax for StringStartsWith?
b) Why we are not supporting LIKE filters in `DataSourceStrategy.translateFilter` ?

Comment (RussellSpitzer) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:241]: Chatted online

a) Requires us using optimization rules in spark since there is a Like Simplifier rule which converts like -> startsWith
b) Is a bug we have in our Spark Code causes unknown filters to crash our pushdown. This comes up whenever a Spark Filter is passed in which doesn't exist in `SparkFilters` I meant to document this last week but ran out of time

Comment (ajantha-bhat) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:241]: Thanks, I have pushed the changes.

**Comment by ajantha-bhat** [comment]
@RussellSpitzer , @jackye1995 , @huaxingao : PR is ready. Please take a look at it again.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java [hunk 9]
```diff
@@ -174,6 +174,82 @@ public void testRewriteDataFilesWithFilterOnPartitionTable() {
...
+
+    // Pass the literal value which is not present in the data files.
+    // So that parsing can be tested on a same dataset without actually compacting the files.
+
+    // EqualTo
+    sql("CALL %s.system.rewrite_data_files(table => '%s'," +
+        " where => 'c1 = 3')", catalogName, tableIdent);
+    // GreaterThan
+    sql("CALL %s.system.rewrite_data_files(table => '%s'," +
+            " where => 'c1 > 3')", catalogName, tableIdent);
```
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:212]: Can you fix indentation in this file?

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java [hunk 10]
```diff
@@ -174,6 +174,82 @@ public void testRewriteDataFilesWithFilterOnPartitionTable() {
...
+  @Test
+  public void testRewriteDataFilesWithAllPossibleFilters() {
+    createPartitionTable();
+    // create 5 files for each partition (c2 = 'foo' and c2 = 'bar')
+    insertData(10);
+
+    // Pass the literal value which is not present in the data files.
+    // So that parsing can be tested on a same dataset without actually compacting the files.
+
+    // EqualTo
```
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:207]: I'm not sure that these comments are helpful, since they just restate the predicate in the WHERE clause.

File: spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala [hunk 11]
```diff
@@ -30,4 +34,18 @@ object SparkExpressionConverter {
     // But these two conversions already exist and well tested. So, we are going with this approach.
     SparkFilters.convert(DataSourceStrategy.translateFilter(sparkExpression, supportNestedPredicatePushdown = true).get)
   }
+
+  @throws[AnalysisException]
+  def collectResolvedSparkExpression(session: SparkSession, tableName: String, where: String): Expression = {
+    var expression:Expression = null
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:40]: Nit: variable names and types should be separated by a space.

Comment (ajantha-bhat) [reply spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala:40]: may be we should check why scala style is not catching these ? It is hard to catch from manual review I guess.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java (hunk 9 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:212]: Also can we check about catching these from checkstyle? Sometimes IDE uses wrong indentation and even setting project code style file is not fixing it.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java (hunk 10 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:207]: I think it is ok. Because it is a text representation and  when multiple filtres exist we will know which one we are focusing on, for example and / or

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java (hunk 9 above)
Comment (rdblue) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java]: I looked into it and it is a large change. We'll get around to it eventually, but for now try updating your IDE settings.

Comment (ajantha-bhat) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java]: ok, I have manually handled and pushed already.

**Review by jackye1995 — APPROVED** [review]
looks good to me, thanks for working on this! @ajantha-bhat please let me know if you would like to have also add fixes in the other version in this PR, or have separated PRs, up to you.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java [hunk 12]
```diff
@@ -251,8 +326,8 @@ private void createPartitionTable() {
   }
 
   private void insertData(int filesCount) {
-    ThreeColumnRecord record1 = new ThreeColumnRecord(1, "foo", "detail1");
-    ThreeColumnRecord record2 = new ThreeColumnRecord(2, "bar", "detail2");
+    ThreeColumnRecord record1 = new ThreeColumnRecord(1, "foo", null);
```
Comment (RussellSpitzer) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:329]: I missed why was this changed?

**Review by RussellSpitzer — APPROVED** [review]
One question about the null's in the test insert, but i'm good for merge as well.

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java (hunk 12 above)
Comment (ajantha-bhat) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:329]: while adding a new test cases for all possible filters, I wanted some null data so that my NOT NULL filter will not execute compaction (doesn't select data). C3 was never used in the testcases. So reused it with null data.

**Comment by ajantha-bhat** [comment]
> looks good to me, thanks for working on this! @ajantha-bhat please let me know if you would like to have also add fixes in the other version in this PR, or have separated PRs, up to you.

I think separated is better as we are following for spark PRs.

**Comment by jackye1995** [comment]
merged, thanks @ajantha-bhat for the fix, and thanks @RussellSpitzer and @rdblue for reviews!
